Update EIP-7805: expand on EL changes (#9381)
ethereum/EIPs: EIPS/eip-7805.md

For each transaction `T` in ILs, perform the following:

1. Check whether `T` is present in `B`. If `T` is present, then jump to the next transaction, else continue with next step.

2. Check whether `B` has enough remaining gas to execute `T`. If `T.gas` > `gas_left`, then jump to the next transaction, else continue with next step.

3. Validate `T` against `S` by checking the nonce and balance of `T.origin`.

  - If `T` is invalid, then continue to the next transaction. 
  
  - If `T` is valid, terminate process and return an `INVALID_INCLUSION_LIST` error.

#### Engine API Changes

We make the following changes to the engine API:

- Add `engine_getInclusionListV1` endpoint to retrieve an IL from the `ExecutionEngine`.
- Add `engine_updatePayloadWithInclusionListV1` endpoint to update a payload with the IL that should be used to build the block. This takes as an argument an 8-byte `payloadId` of the ongoing payload build process, along with the IL itself.
- Modify `engine_newPayload` endpoint to include a parameter for transactions in ILs determined by the IL committee member. If the IL is not satisfied an `INVALID_INCLUSION_LIST` error must be returned.

> The full Engine API changes can be found in the [Execution APIs](https://github.com/ethereum/execution-apis/blob/58b019ea5ee4a84d9767402145b9a88789f0359e/src/engine/experimental/eip7805.md) GitHub repository.

Reviewer: I'm not sure if we can add a link to `execution-apis` but if we can, could you please change the link with this latest one? https://github.com/ethereum/execution-apis/blob/572c54cd9d003f700bb5a57ce201606adbc8fc1c/src/engine/experimental/eip7805.md

### Execution Layer

> On the execution layer, an additional check is introduced for new payloads. After all of the transactions in the payload have been executed, we attempt to execute each valid transaction from ILs that was not present in the payload. If one of those transactions executes successfully, then an error is returned to the CL. Although the block is valid the CL will not attest to it.

Reviewer 2: ```suggestion
On the execution layer, an additional check is introduced for new payloads. After all of the transactions in the payload have been executed, we check whether any transaction from ILs, that is not already present in the payload, could be validly included. In other words, we validate the transaction against the post-state of the block, and also check whether there is sufficient gas left to include it. If one of those transactions could be included, then an error is returned to the CL. Although the block is valid, the CL will not attest to it.
```

Reviewer 3: does it make sense to say "the block is valid" here if we return an error? although the meaning is clear to me, we could clarify a bit. not a blocker for approval though.

a further nit: it's nice to try to keep one sentence per line in the markdown for easier review.
I know that we are inconsistent on that in this EIP though.

Author: @jacobkaufmann I think it's important to specify that the block is valid here, this affects which code path we take in the EL

> Let `B` denote the current block.

Reviewer 2: Let `B` denote the current block.
Let `S` denote the execution state following the execution of the last transaction in `B`.
Let `gas_left` be the gas remaining after execution of `B`.

For each transaction `tx` in ILs, perform the following:

1. Check whether `tx` is present in `B`. If `tx` is present, then jump to the next transaction, else continue with next step.
2. Validate `tx` against `S`.

  - If  `tx` is invalid or `tx.gas > gas_left`, then continue to the next transaction. 
  - If `tx` is valid and `tx.gas <= gas_left`, terminate process and return an error.

Reviewer: I think it's more precise to view `gas_left` as a separate value rather than as a part of the execution state.

---

Let `B` denote the current block.
Let `S` denote the execution state following the execution of the last transaction in `B`.
Let `gas_left` be the gas remaining after execution of `B`.

For each transaction `tx` in ILs, perform the following:

1. Check whether `tx` is present in B. If `tx` is present, then jump to the next transaction, else continue with next step.

2. Check whether `B` has enough remaining gas to execute `tx`. If `tx.gas` > `gas_left`, then jump to the next transaction, else continue with next step.

3. Let `balance` and `nonce` be the balance and nonce of `tx.origin` derived from `S` respectively. If the required funds for `tx` > `balance` or `tx.nonce` < `nonce`, then jump to the next transaction, else terminate process and return an error.

Author: @jihoonsong @fradamt is it sufficient to check the nonce and balance? When we are talking about validating against S are we not simulating execution against S?

Reviewer 4: lgtm

Reviewer 5: I think we may need more checks to make sure that a transaction is includable. As an example, 7702 `SetCodeTransactions` that have empty authorisation lists are invalid.

See [comment](https://github.com/ethereum/execution-specs/pull/1214#discussion_r2102402039).

>   - If `T` is invalid, then continue to the next transaction. 

Reviewer 2: ```suggestion
  - If `T` is invalid against `S`, then continue to the next transaction. 
```

  
  - If `T` is valid, terminate process and return an error.

> 3. Execute `T` on state `S`. Assert that the execution of `T` fails.

Reviewer 2: ```suggestion
```

> If `B` is full, the process terminates. Also note that the state `S` must be restored if a transaction from the IL is successfully executed against it.

Reviewer 2: I'd remove these two lines. I don't think it makes sense to talk about `B` being full, since it's only full if it uses exactly the block gas limit. It makes sense to talk about this tx per tx explicitly, by comparing `tx.gas` and `gas_left` like done above.  I also don't think it makes sense to talk about `S` being restored since we're never doing any execution.

> If `B` is full, the process terminates. Also note that the state `S` must be restored if a transaction from the IL is successfully executed against it.

Reviewer 2: ```suggestion
```

I don't think it makes sense to talk about `B` being full, since it's only full if it uses exactly the block gas limit. It makes sense to talk about this tx per tx explicitly, by comparing `tx.gas` and `gas_left` like done above. 

 I also don't think it makes sense to talk about `S` being restored since we're never doing any execution.

#### Engine API Changes

We make the following changes to the engine API:

- Add `engine_getInclusionListV1` endpoint to retrieve an IL from the `ExecutionEngine`.

> - Add `engine_updatePayloadWithInclusionListV1` endpoint to update a payload with the IL that should be used to build the block. This takes as an argument an 8 byte `payloadId` of the ongoing payload build process, along with the IL itself.

Author: 'bytes' doesn't sound right to me here, but I think there should be a hyphen: https://ell.stackexchange.com/questions/93379/do-you-say-2-byte-or-2-bytes

> - Add `engine_updatePayloadWithInclusionListV1` endpoint to update a payload with the IL that should be used to build the block. This takes as an argument an 8-byte `payloadId` of the ongoing payload build process, along with the IL itself.

Reviewer 6: I'm confused by this. I might be totally getting the CL <-> EL communication wrong though, but I cannot wrap my head around it currently.

`newPayload` has a proposed new parameter for the IL. `newPayload` is used to verify the payload (but does not return a `payloadId`). The `payloadId` is returned by `forkchoiceUpdated`, but only if the CL is proposing a block (then CL will send the extra `payloadAttributes` param to instruct EL to start building). So `newPayload` can be used to verify if the block is valid given the IL, but this can only be done after the build process has finished (`forkchoiceUpdated` with a `payloadAttributes` returns `payloadId`, then `getPayload` to get the payload and then to verify `newPayload` with the IL). This also means that only after the build process has started (via `forkchoiceUpdated`) the `updatePayloadWithInclusionList` can be called (this should interrupt the process and start over by now including the IL). 

I think that these changes should be made instead:

`engine_newPayloadV5`

params: copy from https://github.com/ethereum/execution-apis/blob/3779ff972be60cff5fec93e9b51b0ec24aa52e2d/src/engine/prague.md#engine_newpayloadv4 and include `v. inclusionList` param. This should also verify the `MaxBytesPerInclusionList` check (and spec how this size is measured, for instance the RLP of these txs?)

`engine_forkchoiceUpdatedV4`

Builds on `forkchoiceUpdatedV3` https://github.com/ethereum/execution-apis/blob/3779ff972be60cff5fec93e9b51b0ec24aa52e2d/src/engine/cancun.md#engine_forkchoiceupdatedv3 and adds `PayloadAttributesV4` with the `inclusionList` as extra attribute (this should also verify the `MaxBytesPerInclusionList` I think)

`engine_updatePayloadWithInclusionListV1`

I think this can be removed. To update the IL in the build process, just call `forkchoiceUpdatedV4` again. The `forkchoiceState` will (should?) stay the same in the build process, so it is clear that  the build process should be started over with these new `payloadAttributes` (new IL)

`engine_getInclusionListV1` 

Keep as currently specced

The main point here is that when building the block we should be aware of the IL. Otherwise the verification process is: build the block, then verify it (`newPayload`). The build process returns the `payloadId` and the only way currently to set the IL is via the `updatePayloadWithInclusionList`, but this needs the `payloadId`.

Reviewer: 0.
Regarding `engine_updatePayloadWithInclusionListV1`, the main reason I've added was because I wasn't sure how EL devs would feel about overloading forkchoice attribute. We can remove it and add `engine_forkchoiceUpdatedV4`, if that's what EL devs prefer. We just haven't had enough feedback from EL devs thus far.

Another reason of having `engine_updatePayloadWithInclusionListV1` was that it's easier to implement and we wanted to make faster progress. The extra amount of work of merging it into `engine_forkchoiceUpdatedV4` can be justified given where we're stand at the moment IMHO.

1.
Block building is one thing and block verification is another. A block builder doesn't need to verify their block. They're the one who builds a block according to IL constraints. Attesters are the one who verify a block and attest to it, if it's legit.

Reviewer 6: Could you then show me how the current flow would go for building the block? How would the builder know the IL? I'd assume `forkchoiceUpdated`, and then `updatePayloadWithInclusionList` with the `payloadId` returned from the `forkchoiceUpdated`? This would thus chain 2 calls - right? I'd then really suggest to put this in `forkchoiceUpdated` :thinking:

Reviewer: > I'd assume forkchoiceUpdated, and then updatePayloadWithInclusionList with the payloadId returned from the forkchoiceUpdated?

Yes, this is current implementation.

It's not guaranteed to have all the ILs until 9 seconds into the slot. So if we were to use `forkchoiceUpdatedV4` and make EL call only once, you need to wait until 9 seconds into the slot before initiating the payload building process. I'd like to learn more about how people would think about this change.

I don't have a strong opinion on either way. Thank you for your feedback and I'd encourage more people to give us feedback to find a better solution :)

Reviewer 6: I disagree, the block builder can already start building the block immediately, it can just call `forkchoiceUpdatedV4` and we thus assume the IL list is empty. If the IL list gets updated, then either the `updatePayloadWithInclusionList` could be called (if this is the approach) or if we insert it as param for `forkchoiceUpdated` then it would call it with exactly the same arguments, but with the updated IL.

I would assume that block builders start building the block immediately. I am not sure, but I think it is likely if the node is not censoring that in most cases, if it has received all ILs (or time to receive new ILs is over) then I think it is likely that the block already consists of most of these txs.

But this is a good point, I am not super familiar with the block building process which clients implement, and it would also be worth to study what "big block builders" implement and how they might run into build problems with the new ILs.

Note that the EIP-1559 fee market will target the blocks to be full on average of the `gasTarget` which is half the `gasLimit`. We can thus assume that a "normal block" is 50% full. If the IL then contains transactions which are NOT in the block yet, but are valid (sufficient balance, correct nonce, willing to pay enough for the `baseFee` and enough gas left in the block), then these can be added on top of the existing block (which was already built). On the other hand, if the block is rather full then the txs might not fit in anymore (due to gas limits). But this block is thus per this spec compliant with the ILs and the ELs will accept these blocks (and the CLs will thus attest).

Reviewer: If you add a new engine API, which is `forkchoiceUpdatedV4 ` and make EL call twice, what's the improvement?

Reviewer 6: The freedom that you can start the building process at any moment. The version with `updatePayloadWithInclusionList` makes it mandatory to start the build process with the empty list. Then one has to update/interrupt the build process with the inclusion list by calling `updatePayloadWithInclusionList`

Reviewer: I don't think so. You have the same freedom in `updatePayloadWithInclusionList` approach. In your scenario, the only difference is whether the new engine API is `forkchoiceUpdatedV4` or `updatePayloadWithInclusionList`.

Also, the payload building usually finishes in 100ms. You don't interrupt the building process, you just initiate another round with ILs.

The only case `forkchoiceUpdatedV4` approach is better is when you initiate the payload building process after 11 seconds into the previous slot. Given that most of blocks play timing games, I started to think that this would be better in that sense.

> - Modify `engine_newPayload` endpoint to include a parameter for transactions in ILs determined by the proposer. If the IL is not satisfied an `INVALID_INCLUSION_LIST` error must be returned.

Reviewer 3: ```suggestion
- Modify `engine_newPayload` endpoint to include a parameter for transactions in ILs determined by the validator. If the IL is not satisfied an `INVALID_INCLUSION_LIST` error must be returned.
```

or "attester"? it is not the proposer that supplies the IL that should be satisfied by the proposer's payload.

Reviewer: Or we could use "the IL committee members"

> The full Engine API changes can be found in the [Execution APIs](https://github.com/ethereum/execution-apis/blob/58b019ea5ee4a84d9767402145b9a88789f0359e/src/engine/experimental/eip7805.md) GitHub repository.

Reviewer 3: I don't think we can provide external links.

Reviewer: @jochem-brouwer Updated the execution-api PR, thanks!

Author: @SamWilsn would it be possible to add execution api as an allowed external link?

Author: made a PR to allow links to execution api https://github.com/ethereum/EIPs/pull/9809

#### IL Building

> The rules for building ILs are left up to the implementers discretion. Possible strategies include randomly choosing from the transaction pool, ordering by priority fee, time spent pending in the mempool, etc.

Reviewer: ```suggestion
The rules for building ILs are left to the discretion of implementers. For instance, they may select raw transactions from the public mempool in various ways such as at random, by priority fee, or based on how long they have been pending. This process continues until the IL reaches its maximum size of `MAX_BYTES_PER_INCLUSION_LIST = 8 KiB`.
```

#### Engine API Changes

We make the following changes to the engine API:

- Add `engine_getInclusionListV1` endpoint to retrieve an IL from the `ExecutionEngine`.
- Add `engine_updatePayloadWithInclusionListV1` endpoint to update the IL that should be used to build the block. This takes as an argument an 8 byte `payloadId` of the ongoing payload build process, along with the IL itself.
- Modify `engine_newPayload` endpoint to include a parameter for transactions in ILs determined by the proposer. If the IL is not satisfied an `INVALID_INCLUSION_LIST` error must be returned.

The IL is always encoded as an array of RLP encoded transactions in hex form.

The full Engine API changes can be found in the [Execution APIs](https://github.com/ethereum/execution-apis/blob/58b019ea5ee4a84d9767402145b9a88789f0359e/src/engine/experimental/eip7805.md) GitHub repository.

#### IL Building

> The IL should be built using the following rules:

Reviewer: Actually we don't impose IL ordering rules. Some clients may prefer to pick transactions based on priority fee while others may choose different heuristic algorithms. We expect more diverse approaches on IL construction would foster censorship resistance.

Author: That makes sense, but conflicts with another part of the EIP:

> By default, ILs are built by selecting raw transactions from the public mempool, ordered by priority fees

I can update both places

If `B` is full, the process terminates. Also note that we do not need to reset the state to `S`, since the only way for a transaction to alter the state is for it to execute successfully, in which case the block is invalid, and so the block will not be applied to the state.

#### Engine API Changes

We make the following changes to the engine API:

- Add `engine_getInclusionListV1` endpoint to retrieve an IL from the `ExecutionEngine`.
- Add `engine_updatePayloadWithInclusionListV1` endpoint to update a payload with the IL that should be used to build the block. This takes as an argument an 8 byte `payloadId` of the ongoing payload build process, along with the IL itself.
- Modify `engine_newPayload` endpoint to include a parameter for transactions in ILs determined by the proposer. If the IL is not satisfied an `INVALID_INCLUSION_LIST` error must be returned.

The IL is always encoded as an array of RLP encoded transactions in hex form.

The full Engine API changes can be found in the [Execution APIs](https://github.com/ethereum/execution-apis/blob/58b019ea5ee4a84d9767402145b9a88789f0359e/src/engine/experimental/eip7805.md) GitHub repository.

#### IL Building

> By default, ILs are built by selecting raw transactions from the public mempool, ordered by priority fees, up to the IL’s maximum size in bytes of `MAX_BYTES_PER_INCLUSION_LIST = 8 KiB` per IL. Additional rules can be optionally applied to maximize censorship resistance, such as prioritizing valid transactions that have been pending in the mempool the longest. Implementations are free to choose their own ordering rules.

Reviewer: There is no default rule. While ordering and selecting transactions by priority fee is probably the most anticipated strategy, it should remain as optional, too.

Author: This was moved from elsewhere in the EIP but maybe should be changed

Reviewer: Where did you get it from?

Reviewer: Regarding adding this to the EIP, I want to ask other people's opinions. cc @soispoke @terencechain

Author: previously it was here: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7805.md#il-committee-members

Reviewer: Ah, right, my bad. I prefer to keep it as is and have only engine API changes.

Author: imo it's nice to have it here so the EL changes are contained in this section, easier to know what needs to be implemented. You think having an extra section here is bad for readability?

Reviewer: From the big picture perspective, yes. The original version provides a guidance on how IL committee members could construct their ILs when a reader reads the role of IL committee member. This change removes it and hence brings ambiguity compared to before.

Author: I made a change to mention IL building in that earlier section but then expand on it in the EL section, wdyt?
